### PR TITLE
Add backward support for rudimentary NestedTensor.sum(dim)

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Per offline discussion, this will be updated to use expand once expand semantics for nested tensor have been fleshed out.

Next steps will be to add support for other features for forward sum mentioned on #82387 and likewise update the backward

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82625
* #83356

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -1532,11 +1532,11 @@
   result: auto_linear
 
 - name: sum.SymInt(Tensor self, SymInt[1] dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor
-  self: sum_backward(grad, self.sym_sizes(), dim, keepdim)
+  self: sum_backward_symint(grad, self.sym_sizes(), dim, keepdim)
   result: auto_linear
 
 - name: sum.dim_IntList(Tensor self, int[1]? dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor
```
Comment (mikaylagawarecki) [inline tools/autograd/derivatives.yaml:1538]: @albanD  do you have any thoughts on what is the best way to register a NestedTensor kernel for a backward formula that accepts `self.sizes()`? It seems that this is the case for most of the reductions :( 

Right now I'm doing something super hacky where I add a kwarg self to sum_backward and pass a dummy in place of sizes which is probably really bad. I recall Driss mentioned something about editing VariableTypeManual.cpp but you mentioned yesterday that we probably shouldn't do this, is there any other way around this?

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -1532,11 +1532,11 @@
...
-  self: sum_backward(grad, self.sizes(), dim, keepdim)
+  self: sum_backward(grad, IntArrayRef({-1}), dim, keepdim, self)
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:1539]: The problem is that this is not OK to do. The reason is that if you use `self` here, then the autograd has to save the full Tensor of size `self` and thus this will increase memory usage for all users.

**Comment by albanD** [comment]
This was already discussed in https://docs.google.com/document/d/1dVMH33IK2C7n3Ee8LBtfWw2i8MzxkAIkvuUo_hZY63k/edit#
Any reason for not going with that is proposed there?

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (mikaylagawarecki) [reply tools/autograd/derivatives.yaml:1539]: @albanD completely agree that what I'm doing here is horrible, was just hacking together something to try to get it to compile! If I understand correctly from talking to Driss before, (please correct me if I'm wrong) the scheme in the linked google doc was not implemented for now. I think I heard during the meeting yesterday that it was possible to workaround the `self.sizes()` issue for NestedTensor in VariableTypeManual.cpp though, would you be willing to point me to an example of how this could be done?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -4771,6 +4771,11 @@
 - func: sum.DimnameList_out(Tensor self, Dimname[1] dim, bool keepdim=False, *, ScalarType? dtype=None, Tensor(a!) out) -> Tensor(a!)
   device_check: NoCheck   # TensorIterator
 
+# TODO: this function will be replaced once nested expand semantics have been settled on
+- func: _nested_sum_backward(Tensor grad, Tensor self, int[1]? dim, bool keepdim=False) -> Tensor
+  dispatch:
+    NestedTensorCPU: NestedTensor_sum_dim_backward_CPU
```
Comment (bdhirsh) [inline aten/src/ATen/native/native_functions.yaml:4777]: tiny nit: there isn't really any enforcement about naming conventions in the dispatch table, but maybe name the kernel `_nested_sum_backward_cpu`? (i.e. `{foo}_cpu` for the cpu kernel for an op `{foo}`.

Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:4777]: nit: the usual (from the rest of this file) naming convention for this would be `_nested_sum_backward_cpu`.
I know this is not the current convention for a few nested function but it would be good to align the two no?

File: tools/autograd/derivatives.yaml [hunk 4]
```diff
@@ -1544,8 +1544,13 @@
   result: auto_linear
 
 - name: sum.dim_IntList(Tensor self, int[1]? dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor
-  self: sum_backward(grad, self.sizes(), dim, keepdim)
-  result: auto_linear
+  dispatch:
+    Default:
+      self: sum_backward(grad, self.sizes(), dim, keepdim)
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:1549]: nit: Could you add the following test to TestAutogradMultipleDispatch please?
```python
    def test_sum_not_saving(self, device):
        def foo(x):
            # Don't modify the input inplace
            x = x.clone()
            res = x.sum()
            x.add_(1)
            return res

        inp = torch.rand(2, device=device, requires_grad=True)
        # sum's input is not saved for regular Tensors
        foo(inp).backward()

        # sum's input is saved for nested Tensors
        nt = torch.nested_tensor([torch.rand(2), torch.rand(2)], device=device).requires_grad_()
        with self.assertRaisesRegex(RuntimeError, "has been modified inplace"):
            foo(nt).backward()
```

(I can't run it locally, so there might be a bug there. But something in the same spirit.

File: test/test_nestedtensor.py [hunk 5]
```diff
@@ -616,17 +616,32 @@ def test_nested_tensor_mul_in_place(self, device, dtype):
...
+        def test_sum(device, dtype, ntensors, max_sizes, dim, keepdim=True, requires_grad=False):
+            nt = self.random_nt(device, dtype, ntensors, max_sizes)
             nt2 = nt.clone()
-            nt = nt.sum(dim=dim, keepdim=keepdim)
             ub2 = nt2.unbind()
-            ub2 = [t.sum(-1, keepdim=keepdim) for t in ub2]
-            nt2 = torch.nested_tensor(ub2)
-            self.assertEqual(nt, nt2)
+            if requires_grad:
+                nt.requires_grad_(requires_grad)
```
Comment (albanD) [inline test/test_nestedtensor.py:624]: Why isn't `requires_grad` a kwarg for the `torch.nested_tensor()` constructor? Is that constructor supposed to be differentiable?

File: test/test_nestedtensor.py [hunk 6]
```diff
@@ -616,17 +616,32 @@ def test_nested_tensor_mul_in_place(self, device, dtype):
...
+                [t.requires_grad_(requires_grad) for t in ub2]
+            nt_sum = nt.sum(dim=dim, keepdim=keepdim)
+            ub2_sum = [t.sum(-1, keepdim=keepdim) for t in ub2]
+            self.assertEqual(nt_sum, torch.nested_tensor(ub2_sum))
+            # test backward
+            if requires_grad:
+                # generate gradient tensor that has the same size as the output
```
Comment (albanD) [inline test/test_nestedtensor.py:631]: It would be nice to have `torch.randn_like(nt_sum)` :D

File: test/test_nestedtensor.py [hunk 7]
```diff
@@ -616,17 +616,32 @@ def test_nested_tensor_mul_in_place(self, device, dtype):
...
+                gt = torch.nested_tensor(gt2).clone()
+                nt_sum.backward(gt)
+                for t2, g2 in zip(ub2_sum, gt2):
+                    t2.backward(g2)
+                self.assertEqual(nt.grad, torch.nested_tensor([t.grad for t in ub2]))
             return
 
         for ntensors, max_sizes in params:
-            test_sum(self.random_nt(device, dtype, ntensors, max_sizes), len(max_sizes))
+            test_sum(device, dtype, ntensors, max_sizes, len(max_sizes), requires_grad=True)
```
Comment (albanD) [inline test/test_nestedtensor.py:644]: If requires_grad is always True, no need for it to be a kwarg?

File: test/test_nestedtensor.py (hunk 5 above)
Comment (mikaylagawarecki) [reply test/test_nestedtensor.py:624]: I believe it is supposed to be differentiable, will look into adding this!

File: tools/autograd/derivatives.yaml [hunk 8]
```diff
@@ -1544,8 +1544,13 @@
...
+      result: auto_linear
+    AutogradNestedTensor:
+      # TODO: replace this function once semantics for nested tensor expand have been settled on
+      self: _nested_sum_backward(grad, self, dim, keepdim)
```
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply tools/autograd/derivatives.yaml:1554]: Hm I was under the impression that it was not supported from[ this TORCH_CHECK ](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/forward_grad.cpp#L20) :b

File: test/test_nestedtensor.py (hunk 5 above)
Comment (albanD) [reply test/test_nestedtensor.py:624]: If it is differentiable (which looks like it is), you shouldn't add that.

File: tools/autograd/derivatives.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (albanD) [reply tools/autograd/derivatives.yaml:1554]: Ho wow, well I guess it is not :)

File: test/test_nestedtensor.py (hunk 6 above)
Comment (mikaylagawarecki) [reply test/test_nestedtensor.py]: Could I hold off on this until Driss implements empty_like in #83140, I think that the implementation for randn_like might be similar

Comment (albanD) [reply test/test_nestedtensor.py]: Of course!

btw driss is using `nt = self.random_nt(device, dtype, 4, (4, 4))` in a few of his tests. Maybe that is a good replacement?

In any case, this is a small detail that can be done later.

**Review by albanD — APPROVED** [review]
SGTM
I didn't look into details into the kernel though (but I guess it's fine as it's going to be removed soon).

File: test/test_nestedtensor.py (hunk 6 above)
Comment (albanD) [reply test/test_nestedtensor.py]: Ho you're already using that now :D
